### ISSUE TITLE
[no-unnecessary-type-arguments] Flags a violation for identically-named but different types

### ISSUE THREAD

**Issue by NickHeiner** [typescript-eslint/typescript-eslint#4489]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

[Playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=C4TwDgpgBAIhBmBDArgG2AUSgXigb3gHtCAuKAZ2ACcBLAOwHMBfAbgChRIoAVAHi1xwkaTAD4c+AEaEAHmQys2HcNADiEvkJToMo9mwAmEAMapEVaAFtCBtNADkk8-fxso7qJ2haRA-OUJLCGAAC3oGDFRyCDJqZAhFD08VKHVcTQRtMXYmIA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuBBcSiCAhilFtPFswOaUC2cRBCABcIOChRkUIAL5A&tsConfig=N4XyA)

```TS
type DefaultE = {foo: string};
type T<E = DefaultE> = {box: E};

type G = T<DefaultE>;

declare module 'bar' {
    type DefaultE = {somethingElse: true};
    // This is flagged as an error. The name `DefaultE` is the same, but the type is different.
    type G = T<DefaultE>;
}
```

**Expected Result**
The rule only flags violations when the type argument is identical to the default.

**Actual Result**
The rule flags violations when the type argument is different from the default, but has the same name.

**Comment by bradzacher** [comment]
The rule currently does a name comparison
https://github.com/typescript-eslint/typescript-eslint/blob/d4768c0caa7453b057c70a6a3e1e8248852790b1/packages/eslint-plugin/src/rules/no-unnecessary-type-arguments.ts#L54

It should probably just compare the types by reference instead?
Unsure why it did a text comparison

**Comment by NickHeiner** [comment]
Yeah, by reference seems better. :smile:

**Comment by NickHeiner** [comment]
Thanks!

Are there other TS static analysis libraries you have in mind? I might be interested in using them as well.

**Comment by JoshuaKGoldberg** [comment]
https://github.com/JoshuaKGoldberg/TypeStat is the big one for me, personally. 😄
